csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll

```diff
@@ -3814,150 +3730,212 @@ module ControlFlow {
+        }
 
-      /**
-       * Holds if `succ` with splits `succSplits` is a successor of type `t` for `pred`
-       * with splits `predSplits`.
-       */
-      predicate succSplits(ControlFlowElement pred, Splits predSplits, ControlFlowElement succ, Splits succSplits, SuccessorType t) {
-        exists(Completion c |
-          t.matchesCompletion(c) |
+        private predicate succSplits1(ControlFlowElement pred, Splits predSplits, ControlFlowElement succ, Completion c) {
+          // Case 1a
           exists(Reachability::SameSplitsBlock b |
-            pred = b.getAnElement() |
-            b.isReachable(predSplits) and
-            succ = succ(pred, c) and
+            succReachable(b, pred, predSplits, succ, c) |
             succ = b.getAnElement() and
-            not succ = b and
-            succSplits = predSplits
+            not succ = b
           )
           or
-          succSplits3(pred, predSplits, TSplitsNil(), succ, succSplits, c)
-        )
-      }
+          // Case 1b
+          succSplits1bForall(pred, predSplits, succ, c, TSplitsNil())
+        }
 
-      /**
-       * Holds if `pred` with splits `predSplits` can exit the enclosing callable
-       * `succ` with type `t`.
-       */
-      predicate succExitSplits(ControlFlowElement pred, Splits predSplits, Callable succ, SuccessorType t) {
-        exists(Reachability::SameSplitsBlock b, Completion c |
-          pred = b.getAnElement() |
-          b.isReachable(predSplits) and
-          t.matchesCompletion(c) and
-          succ = succExit(pred, c) and
-          forall(SplitInternal predSplit |
-            predSplit = predSplits.getASplit() |
-            predSplit.hasExit(pred, c)
+        private predicate succSplits2aux(ControlFlowElement pred, Splits predSplits, ControlFlowElement succ, Completion c) {
+          exists(Reachability::SameSplitsBlock b |
+            succReachable(b, pred, predSplits, succ, c) and
+            (succ = b.getAnElement() implies succ = b)
+          |
+            predSplits.getASplit().hasExit(pred, succ, c)
+            or
+            any(SplitInternal split).hasEntry(pred, succ, c)
           )
-        )
+        }
+
+        // A `forall` over the splits in `predSplits`, but written using explicit recursion
+        // to avoid negative recursion compilation error.
+        private predicate succSplits2aNoneOfKindForall(ControlFlowElement pred, Splits predSplits, ControlFlowElement succ, Completion c, SplitKind sk, Splits remaining) {
+          succSplits2aux(pred, predSplits, succ, c) and
+          sk.appliesTo(succ) and
+          remaining = predSplits
+          or
+          exists(Splits mid, SplitInternal split |
+            succSplits2aNoneOfKindForall(pred, predSplits, succ, c, sk, mid) and
+            mid = TSplitsCons(split, remaining)
+          |
+            split.getKind() = any(SplitKind sk0 |
+              sk0 != sk and
+              (sk0.appliesTo(succ) or split.hasExit(pred, succ, c))
```

Comment:
I think the condition `split.hasExit(pred, succ, c)` can be removed as it's already on line 3844.

Reply:
Yes, I have removed it.